content/02_KeyTypes.md

```diff
@@ -1,488 +1,629 @@
+
+Observant readers will have noticed that an example in the preceding chapter looks like it shouldn't work. That code created an `IObservable<long>` that produced events once per second, and then it subscribed to it with this code:
+
+```cs
+ticks.Subscribe(
+    tick => Console.WriteLine($"Tick {tick}"));
+```
+
+That's passing a delegate, and not the `IObserver<T>` that `IObservable<T>.Subscribe` requires. We'll get to `IObserver<T>` shortly, but all that's happening here is that this example is using an extension method from the `System.Reactive` NuGet package:
+
+```cs
+// From the System.Reactive library's ObservableExtensions class
+public static IDisposable Subscribe<T>(this IObservable<T> source, Action<T> onNext)
+```
+
+This is a helper method that wraps a delegate in an implementation of `IObserver<T>` and then passes that to `IObservable<T>.Subscribe`. The effect is that we can write just a simple method (instead of a complete implementation of `IObserver<T>`) and the observable source will invoke our callback each time it wants to supply a value. It's more common to use this kind of helper than to implement Rx's interfaces ourselves.
+
+### Hot and Cold Sources
+
+Since an `IObservable<T>` cannot supply us with values until we subscribe, the time at which we subscribe can be important. Imagine an `IObservable<Trade>` describing trades occurring in some market. If the information it supplies is live, it's not going to tell you about any trades that occurred before you subscribed. In Rx, sources of this kind are described as being _hot_.
+
+Not all sources are _hot_. There's nothing stopping an `IObservable<T>` always supplying the exact same sequence of events to any subscriber no matter when the call to `Subscribe` occurs. (Imagine an `IObservable<Trade>` which, instead of reporting live information, generates notifications based on recorded historical trade data.) Sources where it doesn't matter at all when you subscribe are known as _cold_ sources.
+
+Here are some sources that might be represented as hot observables:
+
+* Measurements from a sensor
+* Price ticks from a trading exchange
+* An event source that distributes events immediately such as Azure Event Grid
+* mouse movements
+* timer events
+* broadcasts like ESB channels or UDP network packets
+
+And some examples of some sources that might make good cold observables:
+
+* the contents of a collection (such as is returned by the [`ToObservable` extension method for `IEnumerable<T>`](03_CreatingObservableSequences.md#from-ienumerablet))
+* a fixed range of values, such as [`Observable.Range`](03_CreatingObservableSequences.md#observablerange) produces
```

Comment:
> 03_CreatingObservableSequences.md#observablerange

Should be 03_CreatingObservableSequences.md#observable.range

Reply:
Should it? VS code thinks it's right as it is. If I add the full stop as you suggest, the link is now broken inside VS Code.

Is this another "opinion is divided on the matter" markdownism?

Reply:
It might be two conflicting approaches.

The Id generated by that heading is https://www.introtorx.com/chapters/creating-observable-sequences#observable.create

Reply:
and is defined by markdig as "auto-identifiers" https://github.com/xoofx/markdig/blob/master/src/Markdig.Tests/Specs/AutoIdentifierSpecs.md